presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/KeyBasedSampler.java

```diff
@@ -0,0 +1,270 @@
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(variableAllocator, "variableAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        if (SystemSessionProperties.isKeyBasedSamplingEnabled(session)) {
+            List<String> sampledFields = new ArrayList<>(2);
+            PlanNode rewritten = SimplePlanRewriter.rewriteWith(new Rewriter(session, metadata.getFunctionAndTypeManager(), idAllocator, sampledFields), plan, null);
+            if (!sampledFields.isEmpty()) {
+                warningCollector.add(new PrestoWarning(SAMPLED_FIELDS, String.format("Sampled the following columns/derived columns at %s percent:\n\t%s", SystemSessionProperties.getKeyBasedSamplingPercentage(session) * 100., String.join("\n\t", sampledFields))));
+            }
+            else {
+                warningCollector.add(new PrestoWarning(SEMANTIC_WARNING, "Sampling could not be performed due to the query structure"));
+            }
+
+            return rewritten;
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator idAllocator;
+        private final List<String> sampledFields;
+
+        private Rewriter(Session session, FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator idAllocator, List<String> sampledFields)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.sampledFields = requireNonNull(sampledFields, "sampledFields is null");
+        }
+
+        private PlanNode addSamplingFilter(PlanNode tableScanNode, Optional<VariableReferenceExpression> rowExpressionOptional, FunctionAndTypeManager functionAndTypeManager)
+        {
+            if (!rowExpressionOptional.isPresent()) {
+                return tableScanNode;
+            }
+            RowExpression rowExpression = rowExpressionOptional.get();
+            RowExpression arg;
+            Type type = rowExpression.getType();
+            if (!Varchars.isVarcharType(type)) {
+                arg = call(
+                        "CAST",
+                        functionAndTypeManager.lookupCast(CAST, rowExpression.getType().getTypeSignature(), VARCHAR.getTypeSignature()),
+                        VARCHAR,
+                        rowExpression);
+            }
+            else {
+                arg = rowExpression;
+            }
+
+            RowExpression sampledArg = call(
```

Comment:
This can throw if the function doesn't exist. You probably want to handle that.

Reply:
Hmm - doesn't  it throw PrestoException? Then we can just propagate it?

Reply:
The error message might not be clear to users. We'd probably throw a function not found error, but the function is not used in user query. Ideally we'd probably want to catch here and rewrite the error message to something like sampling method not found.

Reply:
Done